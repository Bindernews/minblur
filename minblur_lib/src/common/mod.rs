pub mod string_cache;
pub mod tuple_helper;
pub mod macros;
