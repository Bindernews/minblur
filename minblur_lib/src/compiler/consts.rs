use bn_expression::enum_to_from_str;

/// Name of "current instruction" variable
pub const VAR_COUNTER: &str = "counter";
/// Name of function that gives label destination (also quick-expression)
pub const FUNC_LABEL_DEST: &str = "label";
/// Function that returns 1 if a constant is defined
pub const FUNC_DEFINED: &str = "defined";
/// Constant prefix character
pub const CONST_PREFIX: char = '$';

pub const OPTION_LABEL_MODE: &str = "label_mode";

pub const SOURCE_AUTO_LABEL: &str = "<builtin:auto-label>";

/// Enum to help with converting things to and from strings.
///
/// There are macros for easy conversion of one enum to
/// another, so this should help reduce duplication.
pub enum ConstNames {
    Read,
    Write,
    Draw,
    Print,
    DrawFlush,
    PrintFlush,
    GetLink,
    Control,
    Radar,
    Sensor,
    Jump,

    Max,
    Min,
    Angle,
    Len,
    Noise,
    Abs,
    Log,
    Log10,
    Floor,
    Ceil,
    Sqrt,
    Rand,
    Sin,
    Cos,
    Tan,
    Asin,
    Acos,
    Atan,
}

impl ConstNames {
    enum_to_from_str!(
        ConstNames; pub fn name(); pub fn from_name();
        {
            Self::Read => "read",
            Self::Write => "write",
            Self::Draw => "draw",
            Self::Print => "print",
            Self::DrawFlush => "drawflush",
            Self::PrintFlush => "printflush",
            Self::GetLink => "getlink",
            Self::Control => "control",
            Self::Radar => "radar",
            Self::Sensor => "sensor",
            Self::Jump => "jump",
            Self::Max => "max",
            Self::Min => "min",
            Self::Angle => "angle",
            Self::Len => "len",
            Self::Noise => "noise",
            Self::Abs => "abs",
            Self::Log => "log",
            Self::Log10 => "log10",
            Self::Floor => "floor",
            Self::Ceil => "ceil",
            Self::Sqrt => "sqrt",
            Self::Rand => "rand",
            Self::Sin => "sin",
            Self::Cos => "cos",
            Self::Tan => "tan",
            Self::Asin => "asin",
            Self::Acos => "acos",
            Self::Atan => "atan",
        }
    );
}
