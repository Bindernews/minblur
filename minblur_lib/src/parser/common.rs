use std::ops::RangeFrom;

use super::consts::*;
use bn_expression::AValue;
use nom::{
    branch::alt,
    bytes::complete::{escaped, is_not, tag, take_until, take_while, take_while1},
    character::complete::{char, hex_digit1, oct_digit1, one_of},
    combinator::{cut, map, opt, recognize},
    error::context,
    number::complete::double,
    sequence::{delimited, pair, preceded, terminated, tuple},
    AsChar, IResult, InputIter, InputTake, Parser, Slice,
};

pub type Span<'a> = nom_locate::LocatedSpan<&'a str>;
pub type ErrType<'a> = nom::error::VerboseError<Span<'a>>;
pub type MyResult<'a, O> = IResult<Span<'a>, O, ErrType<'a>>;

/// Helper function to convert a span to a string
pub fn span_string(span: &Span) -> String {
    String::from(*span.fragment())
}

/// Combinator that matches any/all spaces, INLCUDING newlines
pub fn sp0_nl(input: Span) -> MyResult<Span> {
    let chars = " \t\r\n";
    take_while(move |c| chars.contains(c))(input)
}

/// Combinator that matches zero or more spaces EXCLUDING newlines
pub fn sp0(input: Span) -> MyResult<Span> {
    let chars = " \t";
    take_while(move |c| chars.contains(c))(input)
}

/// Combinator that matches one or more spaces EXCLUDING newlines
pub fn sp1(input: Span) -> MyResult<Span> {
    let chars = " \t";
    take_while1(move |c| chars.contains(c))(input)
}

pub fn statement_end(input: Span) -> MyResult<Span> {
    alt((nom::character::complete::line_ending, tag(";")))(input)
}

/// Match a decimal integer
fn number_dec(input: Span) -> MyResult<Span> {
    take_while1(move |c: char| c.is_digit(10))(input)
}

/// Match a hex integer
fn number_hex(input: Span) -> MyResult<Span> {
    preceded(tag("0x"), hex_digit1)(input)
}

/// Match an octal integer
fn number_oct(input: Span) -> MyResult<Span> {
    preceded(tag("0"), oct_digit1)(input)
}

#[allow(dead_code)]
/// Match any type of integer
fn number_any_int(input: Span) -> MyResult<Span> {
    alt((number_dec, number_hex, number_oct))(input)
}

/// Extract a decimal integer
fn parse_number_dec(input: Span) -> MyResult<i64> {
    map(number_dec, |s| (*s).parse::<i64>().unwrap())(input)
}

/// Extract a hex integer
fn parse_number_hex(input: Span) -> MyResult<i64> {
    map(number_hex, |s| i64::from_str_radix(*s, 16).unwrap())(input)
}

/// Extract an octal integer
fn parse_number_oct(input: Span) -> MyResult<i64> {
    map(number_oct, |s| i64::from_str_radix(*s, 8).unwrap())(input)
}

/// Extract any integer type
fn parse_number_any_int(input: Span) -> MyResult<i64> {
    alt((parse_number_dec, parse_number_hex, parse_number_oct))(input)
}

pub fn number_any(input: Span) -> MyResult<f64> {
    alt((double, map(parse_number_any_int, |i| i as f64)))(input)
}

/// Parse an escaped string (excluding the surrounding quotes).
fn match_string<'a>(input: Span<'a>) -> MyResult<'a, Span<'a>> {
    let normal = |inp: Span<'a>| {
        take_while1(|c: char| !c.is_control() && !"\\\"\'\r\n\t\0".contains(c)).parse(inp)
    };
    escaped(normal, '\\', one_of("\"nrt\'\\"))(input)
}

/// Parse a string surrounded by quotes, fails if string parsing fails.
pub fn parse_string(input: Span) -> MyResult<Span> {
    context(
        "string",
        preceded(char('\"'), cut(terminated(match_string, char('\"')))),
    )(input)
}

/// Matches a simple ascii name that starts with a letter or _
/// and continues with letters, numbers, and underscores.
pub fn ascii_identifier(input: Span) -> MyResult<Span> {
    context(
        "match name",
        recognize(pair(
            take_while1(|c: char| c.is_ascii_alphabetic() || c == '_'),
            take_while(|c: char| c.is_ascii_alphanumeric() || c == '_'),
        )),
    )
    .parse(input)
}

/// Basic identifier without special characters.
pub fn identifier_mlog(input: Span) -> IResult<Span, Span, ErrType> {
    recognize(tuple((
        opt(char('@')),
        take_while1(|c: char| c.is_alphabetic() || c == '_'),
        take_while(|c: char| c.is_alphanumeric() || c == '_' || c == '.'),
    )))
    .parse(input)
}

/// Recognize a constant (basic identifier prefixed with $)
pub fn identifier_const(input: Span) -> IResult<Span, (char, Span), ErrType> {
    let take_cond = |c: char| (c != '}' && c != '\r' && c != '\n');
    pair(
        char('$'),
        alt((
            delimited(char('{'), take_while1(take_cond), char('}')),
            identifier_mlog,
        )),
    )(input)
}

pub fn match_identifier_basic_const(input: Span) -> IResult<Span, Span, ErrType> {
    alt((recognize(identifier_const), identifier_mlog))(input)
}

pub fn parse_identifier_basic_const(input: Span) -> IResult<Span, String, ErrType> {
    alt((
        map(identifier_const, |(_, name)| format!("${0}", name)),
        map(identifier_mlog, |s| (*s).to_string()),
    ))(input)
}

/// Identifier / variable name with complex names
pub fn identifier(input: Span) -> MyResult<Span> {
    recognize(alt((
        recognize(identifier_const),
        recognize(terminated(identifier_mlog, char('!'))),
        identifier_mlog,
    )))(input)
}

pub fn _match_value(input: Span) -> MyResult<Span> {
    alt((parse_string, identifier_mlog, number_any_int))(input)
}

/// Parse an expression, either a variable, goto, number, or constant identifier
pub fn parse_value(input: Span) -> MyResult<AValue> {
    alt((
        map(parse_string, |s| AValue::string(*s)),
        map(parse_identifier_basic_const, AValue::name),
        map(number_any, AValue::Num),
    ))(input)
}

/// Match a pair of balanced `open` and `close` characters, returns the entire contents as a string.
///
pub fn balanced_braces<I, E>(open: char, close: char) -> impl FnMut(I) -> nom::IResult<I, I, E>
where
    E: nom::error::ParseError<I>,
    I: Slice<RangeFrom<usize>> + InputIter + InputTake,
    <I as InputIter>::Item: AsChar,
{
    use nom::Needed;

    let balanced = move |input: I| {
        let mut depth = 1;
        for (i, elem) in input.iter_indices() {
            let c = elem.as_char();
            if c == open {
                depth += 1;
            }
            if c == close {
                depth -= 1;
            }
            if depth == 0 {
                return Ok(input.take_split(i));
            }
        }
        Err(nom::Err::Incomplete(Needed::Unknown))
    };
    move |input: I| -> nom::IResult<I, I, E> { delimited(char(open), balanced, char(close))(input) }
}

/// Match a comment to the end of the line
pub fn eol_comment(input: Span) -> MyResult<Span> {
    recognize(pair(char(COMMENT_CHAR), is_not("\n\r")))(input)
}

/// Match a block comment. Does NOT support nested comments.
pub fn block_comment(input: Span) -> MyResult<Span> {
    recognize(tuple((
        tag(COMMENT_BEGIN),
        take_until(COMMENT_END),
        tag(COMMENT_END),
    )))(input)
}

/// Returns a nom error if the input string isn't empty
///
/// Param `f` should generate an approriate error given the input
pub fn assert_input_consumed<'a, F, O>(mut f: F) -> impl FnMut(Span<'a>) -> MyResult<'a, ()>
where
    F: Parser<Span<'a>, O, ErrType<'a>>,
{
    move |input| {
        if (*input).is_empty() {
            Ok((input, ()))
        } else {
            let (input, _) = f.parse(input)?;
            Ok((input, ()))
        }
    }
}
