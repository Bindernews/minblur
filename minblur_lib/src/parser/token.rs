use instruction::{parse_instruction_arg_string, InstructionParser};
use nom::{
    branch::alt,
    bytes::complete::{is_not, tag, tag_no_case, take_until},
    character::complete::{char, one_of},
    combinator::{cut, eof, map, peek, recognize, value},
    error::context,
    multi::{many1, separated_list0},
    sequence::{delimited, pair, preceded, separated_pair, terminated, tuple},
    Finish, Needed, Parser, Slice,
};

use super::{
    common::{
        assert_input_consumed, identifier, identifier_basic, parse_string, sp0, sp0_nl, sp1,
        span_string, ErrType, MyResult, Span,
    },
    consts::*,
    statement::*,
};
use crate::common::{macros::prcall, string_cache::*};
use crate::compiler::instruction::{InstValue, Instruction};

/// Match a comment to the end of the line and return ()
fn eol_comment(input: Span) -> MyResult<Span> {
    recognize(pair(char(COMMENT_CHAR), is_not("\n\r")))(input)
}

/// Match a block comment. Does NOT support nested comments.
fn block_comment(input: Span) -> MyResult<Span> {
    recognize(tuple((
        tag(COMMENT_BEGIN),
        take_until(COMMENT_END),
        tag(COMMENT_END),
    )))(input)
}

/// Match a pair of balanced `open` and `close` characters, returns the entire contents as a string.
///
fn balanced_braces<'a>(open: char, close: char) -> impl FnMut(Span<'a>) -> MyResult<Span<'a>> {
    let balanced = move |input: Span<'a>| -> MyResult<Span<'a>> {
        let mut depth = 1;
        for (i, c) in input.chars().enumerate() {
            if c == open {
                depth += 1;
            }
            if c == close {
                depth -= 1;
            }
            if depth == 0 {
                let remain = input.slice(i..);
                let result = input.slice(0..i);
                return Ok((remain, result));
            }
        }
        Err(nom::Err::<ErrType>::Incomplete(Needed::Unknown))
    };
    delimited(char(open), balanced, char(close))
}

/// Match pair of common balanced braces, returning the contents
///
fn balanced_braces_common(input: Span) -> MyResult<Span> {
    alt((
        balanced_braces('{', '}'),
        balanced_braces('(', ')'),
        balanced_braces('[', ']'),
    ))(input)
}

/// Parse an argument list (arg1, arg2, arg3...)
fn argument_list(input: Span) -> MyResult<Vec<Span>> {
    delimited(
        char('('),
        separated_list0(char(','), delimited(sp0_nl, identifier, sp0_nl)),
        char(')'),
    )(input)
}

/// Parse a condition statement that lasts until the end of the line
fn condition_eol_or_comment(input: Span) -> MyResult<Span> {
    is_not("\r\n#")(input)
}

/// Parse until the end of the line, or a line comment
/// TODO: allow backslash to escape EOL
fn until_eol_or_comment(input: Span) -> MyResult<Span> {
    is_not("\r\n#")(input)
}

/// Parse until either a comment, semicolon, or newline
fn until_end_of_statement(input: Span) -> MyResult<Span> {
    terminated(
        sp0,
        peek(alt((value((), one_of(";#\r\n")), value((), eof)))),
    )(input)
}

#[derive(Clone, Debug, PartialEq, Hash)]
pub struct Token {
    pub line: u32,
    pub column: u32,
    pub data: TokenData,
}
impl Token {
    pub fn new(line: u32, column: u32, data: TokenData) -> Self {
        Self { line, column, data }
    }
}

#[derive(Clone, Debug, PartialEq, Hash)]
pub enum TokenData {
    DirectiveToken(DirectiveToken),
    Instruction(Instruction),
    Label(Label),
    MacroCall {
        name: CaString,
        content: String,
    },
    OutputComment(String),
    /// Comment, EOL, etc.
    Skip,
    /// End of a statement (newline or ;)
    EndOfStatement,
}
#[derive(Clone, Debug, PartialEq, Hash)]
pub enum DirectiveToken {
    Define { key: String, value: InstValue },
    Include { path: String },
    LoadExtension { ext_type: String, path: String },
    InlineExtension { ext_type: String, code: String },
    BeginMacro { name: String, args: Vec<String> },
    EndMacro,
    Option { key: String, value: String },
    If { cond: String },
    Else,
    ElseIf { cond: String },
    EndIf,
}

pub struct TokenParser {
    pub string_cache: StringCache,
    instr_parser: InstructionParser,
}
impl TokenParser {
    pub fn new(string_cache: &StringCache) -> Self {
        Self {
            string_cache: string_cache.clone(),
            instr_parser: InstructionParser::new(string_cache),
        }
    }
}
impl<'a> TokenParser {
    fn parse_token_instruction(&self, input: Span<'a>) -> MyResult<'a, TokenData> {
        map(
            terminated(
                context("instruction", prcall!(self.instr_parser.parse_instruction)),
                until_end_of_statement,
            ),
            TokenData::Instruction,
        )(input)
    }

    fn parse_token_macro_call(&self, input: Span<'a>) -> MyResult<'a, TokenData> {
        fn trim_right_mut(mut s: String) -> String {
            s.truncate(s.trim_end().len());
            s
        }

        map(
            pair(
                terminated(identifier_basic, char('!')),
                alt((
                    preceded(sp0, balanced_braces_common),
                    preceded(sp1, until_eol_or_comment),
                )),
            ),
            |f| TokenData::MacroCall {
                name: self.string_cache.get(*f.0),
                content: trim_right_mut(span_string(&f.1)),
            },
        )(input)
    }

    fn parse_token_label(input: Span) -> MyResult<TokenData> {
        map(terminated(identifier_basic, char(':')), |f| {
            TokenData::Label(Label::new(*f))
        })
        .parse(input)
    }

    fn parse_token_output_comment(input: Span) -> MyResult<TokenData> {
        preceded(peek(tag("#!")), eol_comment)
            .map(|s| TokenData::OutputComment(s.to_string()))
            .parse(input)
    }

    fn parse_token_skip(input: Span) -> MyResult<TokenData> {
        map(alt((sp1, eol_comment, block_comment)), |_| TokenData::Skip)(input)
    }

    fn parse_token_end_of_statement(input: Span) -> MyResult<TokenData> {
        map(one_of(";\r\n"), |_| TokenData::EndOfStatement)(input)
    }

    /// Parse a directive. Directives begin with "." and
    pub fn parse_directive_token(&self, input: Span<'a>) -> MyResult<'a, TokenData> {
        let map_define = tuple((
            value((), pair(tag_no_case(DIRECTIVE_DEFINE), sp1)),
            identifier_basic,
            value((), sp1),
            |inp| parse_instruction_arg_string(&self.string_cache, inp),
        ))
        .map(|opt: ((), Span, (), InstValue)| DirectiveToken::Define {
            key: span_string(&opt.1),
            value: opt.3,
        });

        let parse_include = preceded(preceded(tag_no_case(DIRECTIVE_INCLUDE), sp1), parse_string);
        let map_include = map(parse_include, |s| DirectiveToken::Include {
            path: span_string(&s),
        });

        let parse_macro = preceded(
            preceded(tag_no_case(DIRECTIVE_MACRO), sp1),
            tuple((identifier_basic, sp0, argument_list)),
        );
        let map_macro = map(parse_macro, |tup: (Span, Span, Vec<Span>)| {
            DirectiveToken::BeginMacro {
                name: span_string(&tup.0),
                args: tup.2.iter().map(span_string).collect(),
            }
        });

        let map_end_macro = value(DirectiveToken::EndMacro, tag_no_case(DIRECTIVE_ENDMACRO));

        let parse_option = preceded(
            preceded(tag_no_case(DIRECTIVE_OPTION), sp1),
            separated_pair(identifier_basic, sp1, parse_string),
        );
        let map_option = map(parse_option, |opt| DirectiveToken::Option {
            key: span_string(&opt.0),
            value: span_string(&opt.1),
        });

        let parse_if = preceded(
            preceded(tag_no_case(DIRECTIVE_IF), sp1),
            condition_eol_or_comment,
        );
        let map_if = map(parse_if, |f| DirectiveToken::If {
            cond: span_string(&f),
        });

        let map_else = value(DirectiveToken::Else, tag_no_case(DIRECTIVE_ELSE));

        let map_elseif = map(
            preceded(
                preceded(tag_no_case(DIRECTIVE_ELSEIF), sp1),
                condition_eol_or_comment,
            ),
            |cond| DirectiveToken::ElseIf {
                cond: span_string(&cond),
            },
        );

        let map_endif = value(DirectiveToken::EndIf, tag_no_case(DIRECTIVE_ENDIF));

        let alt_directives = alt((
            map_define,
            map_include,
            map_macro,
            map_end_macro,
            map_option,
            map_if,
            map_else,
            map_elseif,
            map_endif,
        ));

        preceded(
            char(DIRECTIVE_PREFIX),
            context("invalid directive", cut(alt_directives)),
        )
        .map(TokenData::DirectiveToken)
        .parse(input)
    }

    /// Parse a single line (or multi-line block in some cases)
    pub fn parse_next_token(&self, input: Span<'a>) -> MyResult<'a, Token> {
        context(
            "expected label, instruction, or directive",
            alt((
                prcall!(self.parse_directive_token),
                Self::parse_token_label,
                prcall!(self.parse_token_macro_call),
                prcall!(self.parse_token_instruction),
                Self::parse_token_output_comment,
                Self::parse_token_skip,
                Self::parse_token_end_of_statement,
            )),
        )
        .map(|tok: TokenData| Token::new(input.location_line(), input.get_column() as u32, tok))
        .parse(input)
    }

    /// Parse and tokenize an entire assembly file.
    pub fn parse_all_tokens(&self, input: Span<'a>) -> MyResult<'a, Vec<Token>> {
        let (input, output) = many1(prcall!(self.parse_next_token))(input)?;
        let (input, _) = assert_input_consumed(prcall!(self.parse_next_token)).parse(input)?;
        Ok((input, output))
    }

    pub fn parse_str_input(&self, input: &'a str) -> Result<Vec<Token>, ErrType<'a>> {
        self.parse_all_tokens(Span::new(input))
            .finish()
            .map(|out| out.1)
    }
}

pub mod instruction {
    use bn_expression::basic::BasicOp;
    use bn_expression::parse::{match_name_basic, ExpressionParser, IExpressionParser};
    use nom::{
        branch::alt,
        bytes::complete::{tag, take_while1},
        character::complete::char,
        combinator::{cut, map},
        error::{context, make_error, ErrorKind as NomErrorKind},
        sequence::{delimited, preceded, separated_pair},
        InputTakeAtPosition, Parser,
    };

    use crate::common::macros::prcall;
    use crate::common::string_cache::StringCache;
    use crate::common::tuple_helper::TupleToArray;
    use crate::compiler::instruction::*;
    use crate::parser::common::{
        identifier_basic, parse_value, parse_value_no_string, sp1, ErrType, MyResult, Span,
    };

    // Since the macro is local-only we don't have to worry about full-use-paths.
    macro_rules! fn_parse_instr {
        (
            $self:ident;
            $fn_name:ident,
            $map_type:ty,
            $name:literal,
            ( $($arg:ident),+ $(,)? )
        ) => {
            pub fn $fn_name (&$self, input: Span<'a>) -> MyResult<'a, $map_type> {
                preceded(
                    tag($name),
                    map(
                        nom::sequence::tuple(( $(prcall!($self.$arg), )+ )),
                        |v| <$map_type>::from(v.to_array()),
                    )
                )(input)
            }
        };
    }

    #[derive(Debug, Clone, PartialEq)]
    pub struct InstructionParser {
        string_cache: StringCache,
    }
    impl InstructionParser {
        pub fn new(cache: &StringCache) -> Self {
            Self {
                string_cache: cache.clone(),
            }
        }
    }
    impl<'a> Parser<Span<'a>, Instruction, ErrType<'a>> for InstructionParser {
        fn parse(&mut self, input: Span<'a>) -> MyResult<'a, Instruction> {
            self.parse_instruction(input)
        }
    }
    impl<'a> InstructionParser {
        fn_parse_instr!(self; parse_i_read, InstructionRead, "read", 
            ( next_arg, next_arg, next_arg ) );
        fn_parse_instr!(self; parse_i_write, InstructionWrite, "write",
            ( next_arg, next_arg, next_arg ) );
        fn_parse_instr!(self; parse_i_draw, InstructionDraw, "draw",
            ( parse_enum_arg, next_arg, next_arg, next_arg, next_arg, next_arg, next_arg )
        );
        fn_parse_instr!(self; parse_i_print, InstructionPrint, "print",
            ( next_arg_string ) );
        fn_parse_instr!(self; parse_i_draw_flush, InstructionDrawFlush, "drawflush",
            ( next_arg ) );
        fn_parse_instr!(self; parse_i_print_flush, InstructionPrintFlush, "printflush",
            ( next_arg ) );
        fn_parse_instr!(self; parse_i_get_link, InstructionGetLink, "getlink",
            ( next_arg, next_arg ) );
        fn_parse_instr!(self; parse_i_control, InstructionControl, "control",
            ( parse_enum_arg, next_arg, next_arg, next_arg, next_arg, next_arg ) );
        fn_parse_instr!(self; parse_i_radar, InstructionRadar, "radar",
            (
                parse_enum_arg, parse_enum_arg, parse_enum_arg,
                parse_enum_arg, next_arg, next_arg, next_arg,
            )
        );
        fn_parse_instr!(self; parse_i_sensor, InstructionSensor, "sensor",
            ( next_arg, next_arg, next_arg ) );
        fn_parse_instr!(self; parse_i_set, InstructionSet, "set",
            ( next_arg, next_arg ) );
        fn_parse_instr!(self; parse_i_op, InstructionOp, "op",
            ( parse_enum_arg, next_arg, next_arg, next_arg ) );
        pub fn parse_i_end(&self, input: Span<'a>) -> MyResult<'a, InstructionEnd> {
            map(tag("end"), |_| InstructionEnd {})(input)
        }
        fn_parse_instr!(self; parse_i_jump, InstructionJump, "jump",
            ( next_arg, parse_enum_arg, next_arg, next_arg ) );
        fn_parse_instr!(self; parse_i_unit_bind, InstructionUnitBind, "ubind",
            ( next_arg ) );
        fn_parse_instr!(self; parse_i_unit_control, InstructionUnitControl, "ucontrol",
            ( parse_enum_arg, next_arg, next_arg, next_arg, next_arg, next_arg ) );
        fn_parse_instr!(self; parse_i_unit_radar, InstructionUnitRadar, "uradar",
            (
                parse_enum_arg, parse_enum_arg, parse_enum_arg,
                parse_enum_arg, next_arg, next_arg, next_arg,
            )
        );
        fn_parse_instr!(self; parse_i_unit_locate, InstructionUnitLocate, "ulocate",
            (
                parse_enum_arg, parse_enum_arg,
                next_arg, next_arg, next_arg, next_arg, next_arg, next_arg,
            )
        );

        pub fn parse_instruction(&self, input: Span<'a>) -> MyResult<'a, Instruction> {
            alt((
                map(prcall!(self.parse_i_read), Into::into),
                map(prcall!(self.parse_i_write), Into::into),
                // These are before print and draw to make sure the longer names are properly detected
                map(prcall!(self.parse_i_draw_flush), Into::into),
                map(prcall!(self.parse_i_print_flush), Into::into),
                map(prcall!(self.parse_i_print), Into::into),
                map(prcall!(self.parse_i_draw), Into::into),
                map(prcall!(self.parse_i_get_link), Into::into),
                map(prcall!(self.parse_i_control), Into::into),
                map(prcall!(self.parse_i_radar), Into::into),
                map(prcall!(self.parse_i_sensor), Into::into),
                map(prcall!(self.parse_i_set), Into::into),
                map(prcall!(self.parse_i_op), Into::into),
                map(prcall!(self.parse_i_end), Into::into),
                map(prcall!(self.parse_i_jump), Into::into),
                map(prcall!(self.parse_i_unit_bind), Into::into),
                map(prcall!(self.parse_i_unit_control), Into::into),
                map(prcall!(self.parse_i_unit_radar), Into::into),
                map(prcall!(self.parse_i_unit_locate), Into::into),
            ))(input)
        }

        /// Wrap an enum to parse either a constant or name
        pub fn parse_enum_arg(&self, input: Span<'a>) -> MyResult<'a, InstValue> {
            preceded(
                sp1,
                alt((
                    |inp| parse_instruction_value_const(&self.string_cache, inp),
                    map(match_name, |s| {
                        InstValue::EnumName(self.string_cache.get(*s))
                    }),
                )),
            )(input)
        }

        fn next_arg(&self, input: Span<'a>) -> MyResult<'a, InstValue> {
            preceded(
                sp1,
                alt((
                    |inp| parse_instruction_value_const(&self.string_cache, inp),
                    map(parse_value_no_string, InstValue::Value),
                )),
            )(input)
        }

        fn next_arg_string(&self, input: Span<'a>) -> MyResult<'a, InstValue> {
            preceded(
                sp1,
                alt((
                    |inp| parse_instruction_value_const(&self.string_cache, inp),
                    map(parse_value, InstValue::Value),
                )),
            )(input)
        }
    }

    //
    // Helper functions
    //

    /// Match an instruction name
    pub fn match_name(input: Span) -> MyResult<Span> {
        input.split_at_position1_complete(
            |c| !(c.is_alphanumeric() || c == '_'),
            NomErrorKind::AlphaNumeric,
        )
    }

    pub fn parse_instruction_value_const<'a>(
        string_cache: &'_ StringCache,
        input: Span<'a>,
    ) -> MyResult<'a, InstValue> {
        let parse_const_expression = |input| -> MyResult<'a, InstValue> {
            let take_cond = |c: char| (c != '}' && c != '\r' && c != '\n');
            delimited(char('{'), take_while1(take_cond), cut(char('}')))
                .and_then(|input| {
                    // TODO: better error conversion, maybe custom error type?
                    let expr = new_expr_parser()
                        .parse_expression(input)
                        .map_err(nom::Err::Failure)?;
                    Ok(("".into(), InstValue::new_expr(expr)))
                })
                .parse(input)
        };

        let parse_quick_const = map(
            separated_pair(match_name, tag(":"), identifier_basic),
            |(key, value)| InstValue::QuickConstExpr(string_cache.get(*key), (*value).into()),
        );

        preceded(
            char('$'),
            alt((
                // Const in ${}
                context("const expression", parse_const_expression),
                // "quick const-expr" <quick-name> ":" <quick-value>
                context("quick-const expression", parse_quick_const),
                // Simple const name
                map(identifier_basic, |v| InstValue::Const((*v).into())),
                // Error b/c we have a $ but we failed to parse a const or constexpr
                |inp| Err(nom::Err::Error(make_error(inp, NomErrorKind::Fail))),
            )),
        )
        .parse(input)
    }

    fn new_expr_parser() -> impl IExpressionParser<BasicOp> {
        ExpressionParser::<BasicOp, _, _>::new(match_name_basic, match_name_basic)
    }

    pub fn parse_instruction_arg_string<'a>(
        string_cache: &'_ StringCache,
        input: Span<'a>,
    ) -> MyResult<'a, InstValue> {
        alt((
            |inp| parse_instruction_value_const(string_cache, inp),
            map(parse_value, InstValue::Value),
        ))
        .parse(input)
    }
}

#[cfg(test)]
pub mod test {
    use insta;
    use nom::Finish;

    use super::*;

    fn parse_into_tokens<'a>(
        string_cache: &StringCache,
        inp: &'a str,
    ) -> Result<(Span<'a>, Vec<Token>), ErrType<'a>> {
        TokenParser::new(string_cache)
            .parse_all_tokens(Span::new(inp))
            .finish()
    }

    #[test]
    fn directive_macro_begin() {
        let string_cache = StringCache::new();
        let inp = ".macro abc(arg1, arg2 ) ";
        let out = parse_into_tokens(&string_cache, inp).unwrap().1;
        insta::assert_debug_snapshot!(out);
    }

    #[test]
    fn multiline_basic() {
        let string_cache = StringCache::new();
        let inp = "start: \n  set r1 0\n  op add r2 r1 5\n  m! x = r2 * 3\n  op mul y r1 5\n";
        let out = parse_into_tokens(&string_cache, inp).unwrap().1;
        insta::assert_debug_snapshot!(out);
    }
}
