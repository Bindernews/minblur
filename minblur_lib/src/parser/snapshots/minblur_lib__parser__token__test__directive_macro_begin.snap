---
source: minblur_lib/src/parser/token.rs
assertion_line: 334
expression: out

---
[
    Token {
        line: 1,
        column: 1,
        data: DirectiveToken(
            BeginMacro {
                name: "abc",
                args: [
                    "arg1",
                    "arg2",
                ],
            },
        ),
    },
    Token {
        line: 1,
        column: 24,
        data: Skip,
    },
]
